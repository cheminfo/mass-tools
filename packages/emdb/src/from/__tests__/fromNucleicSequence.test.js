/* eslint-disable import/no-extraneous-dependencies */

const { toBeDeepCloseTo } = require('jest-matcher-deep-close-to');

expect.extend({ toBeDeepCloseTo });

const DBManager = require('../..');

describe('fromNucleicSequence', () => {
  it('fromNucleicSequence', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('AAA', {
      ionizations: 'H+,Na+',
      fragmentation: {
        a: true,
      },
      info: {
        kind: 'dna',
      },
    });
    let nucleic = dbManager.databases.nucleic.sort((a, b) => a.ms.em - b.ms.em);
    expect(nucleic).toHaveLength(6);
    expect(nucleic[0]).toBeDeepCloseTo({
      charge: 0,
      em: 313.05760550518,
      mw: 313.2069506932622,
      ionization: { mf: 'H+', em: 1.00782503223, charge: 1, atoms: { H: 1 } },
      unsaturation: 8,
      atoms: { C: 10, H: 12, N: 5, O: 5, P: 1 },
      ms: { ionization: 'H+', em: 314.0648819575009, charge: 1 },
      parts: ['HODampO-1H-1'],
      mf: 'C10H12N5O5P',
      sequence: 'A',
      comment: 'a1',
    });
  });

  it('fromNucleicSequence ds-DNA', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('AAA', {
      ionizations: 'H+,Na+',
      fragmentation: {
        a: true,
      },
      info: {
        kind: 'dsdna',
      },
    });

    let nucleic = dbManager.databases.nucleic.sort((a, b) => a.ms.em - b.ms.em);

    expect(nucleic).toHaveLength(12);
    expect(nucleic).toMatchSnapshot();
  });

  it('TACGTGCCAATAC internal fragment', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('TACGTGCCAATAC', {
      ionizations: '(H+)-5',
      fragmentation: {
        abw: true,
      },
      info: {
        kind: 'dna',
      },
    });

    let nucleic = dbManager.databases.nucleic.sort((a, b) => {
      if (a.ms.em !== b.ms.em) return a.ms.em - b.ms.em;
      if (a.ms.comment < b.ms.comment) return 1;
      return -1;
    });
    expect(nucleic).toHaveLength(56);

    expect(nucleic[0]).toBeDeepCloseTo({
      charge: 0,
      em: 467.04948243777,
      mw: 467.2622646800317,
      ionization: {
        mf: '(H+)-5',
        em: -5.03912516115,
        charge: -5,
        atoms: { H: -5 },
      },
      unsaturation: 8,
      atoms: { C: 14, H: 19, N: 3, O: 11, P: 2 },
      ms: { ionization: '(H+)-5', em: 92.40262003523306, charge: -5 },
      parts: ['HODcmpFurp'],
      mf: 'C14H19N3O11P2',
      sequence: 'Câ¬ ',
      comment: 'w6:a9-B',
    });
  });

  it('TACGTGCCAATAC callback', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('TACGTGCCAATAC', {
      ionizations: '(H+)-5',
      fragmentation: {
        abw: true,
      },
      info: {
        kind: 'dna',
      },
      filter: { callback: (entry) => entry.unsaturation % 2 === 0 },
    });

    let nucleic = dbManager.databases.nucleic.sort((a, b) => {
      if (a.ms.em !== b.ms.em) return a.ms.em - b.ms.em;
      if (a.ms.comment < b.ms.comment) return 1;
      return -1;
    });
    expect(nucleic).toHaveLength(30);
  });

  it('AGGCAG fragment', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('AGGCAG', {
      ionizations: '(H+)-',
      fragmentation: {
        y: true,
      },
      info: {
        kind: 'dna',
      },
    });
    let nucleic = dbManager.databases.nucleic.sort((a, b) => a.ms.em - b.ms.em);
    expect(nucleic).toHaveLength(6);
    expect(nucleic).toMatchSnapshot();
  });

  it('AGG with d-h2o and base loss', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('AGG', {
      ionizations: '(H+)-1',
      fragmentation: {
        dh2o: true,
        baseLoss: true,
      },

      info: {
        kind: 'dna',
      },
    });
    let nucleic = dbManager.databases.nucleic.sort((a, b) => a.ms.em - b.ms.em);
    expect(nucleic).toHaveLength(5);
    expect(nucleic).toMatchSnapshot();
  });

  it('HODamDamDamDamDamH', async () => {
    let dbManager = new DBManager();
    await dbManager.fromNucleicSequence('HODamDamDamDamDamH', {
      ionizations: '(H+)-',
      fragmentation: {
        y: true,
      },
      info: {
        kind: 'dna',
      },
    });
    let nucleic = dbManager.databases.nucleic.sort((a, b) => a.ms.em - b.ms.em);
    expect(nucleic).toHaveLength(5);
    expect(nucleic).toMatchSnapshot();
  });
});
