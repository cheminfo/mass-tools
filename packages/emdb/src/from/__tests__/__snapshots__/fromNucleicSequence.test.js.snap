// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromNucleicSequence AGG with d-h2o and base loss 1`] = `
[
  {
    "atoms": {
      "C": 10,
      "H": 13,
      "N": 5,
      "O": 8,
      "P": 2,
    },
    "charge": 0,
    "comment": "d1-H2O",
    "em": 393.02393639454,
    "ionization": {
      "atoms": {
        "H": -1,
      },
      "charge": -1,
      "em": -1.00782503223,
      "mf": "(H+)-1",
    },
    "mf": "C10H13N5O8P2",
    "ms": {
      "charge": -1,
      "em": 392.01665994221906,
      "ionization": "(H+)-1",
    },
    "mw": 393.1868682186928,
    "parts": [
      "HODampPO2",
    ],
    "sequence": "A",
    "unsaturation": 8,
  },
  {
    "atoms": {
      "C": 20,
      "H": 25,
      "N": 10,
      "O": 14,
      "P": 3,
    },
    "charge": 0,
    "comment": "d2-H2O",
    "em": 722.07645651929,
    "ionization": {
      "atoms": {
        "H": -1,
      },
      "charge": -1,
      "em": -1.00782503223,
      "mf": "(H+)-1",
    },
    "mf": "C20H25N10O14P3",
    "ms": {
      "charge": -1,
      "em": 721.069180066969,
      "ionization": "(H+)-1",
    },
    "mw": 722.3932238362731,
    "parts": [
      "HODampDgmpPO2",
    ],
    "sequence": "AG",
    "unsaturation": 15,
  },
  {
    "atoms": {
      "C": 25,
      "H": 33,
      "N": 10,
      "O": 17,
      "P": 3,
    },
    "charge": 0,
    "comment": "G*",
    "em": 838.1238006358399,
    "ionization": {
      "atoms": {
        "H": -1,
      },
      "charge": -1,
      "em": -1.00782503223,
      "mf": "(H+)-1",
    },
    "mf": "C25H33N10O17P3",
    "ms": {
      "charge": -1,
      "em": 837.1165241835189,
      "ionization": "(H+)-1",
    },
    "mw": 838.5086441253504,
    "parts": [
      "HODampDgmpDgmpH(C-5H-5N-5O-1)",
    ],
    "sequence": "AGG",
    "unsaturation": 16,
  },
  {
    "atoms": {
      "C": 25,
      "H": 33,
      "N": 10,
      "O": 18,
      "P": 3,
    },
    "charge": 0,
    "comment": "A*",
    "em": 854.1187152554099,
    "ionization": {
      "atoms": {
        "H": -1,
      },
      "charge": -1,
      "em": -1.00782503223,
      "mf": "(H+)-1",
    },
    "mf": "C25H33N10O18P3",
    "ms": {
      "charge": -1,
      "em": 853.111438803089,
      "ionization": "(H+)-1",
    },
    "mw": 854.5080490496687,
    "parts": [
      "HODampDgmpDgmpH(C-5H-5N-5)",
    ],
    "sequence": "AGG",
    "unsaturation": 16,
  },
  {
    "atoms": {
      "C": 30,
      "H": 38,
      "N": 15,
      "O": 18,
      "P": 3,
    },
    "charge": 0,
    "em": 989.1732104387099,
    "ionization": {
      "atoms": {
        "H": -1,
      },
      "charge": -1,
      "em": -1.00782503223,
      "mf": "(H+)-1",
    },
    "mf": "C30H38N15O18P3",
    "ms": {
      "charge": -1,
      "em": 988.165933986389,
      "ionization": "(H+)-1",
    },
    "mw": 989.6349483608528,
    "parts": [
      "HODampDgmpDgmpH",
    ],
    "sequence": "AGG",
    "unsaturation": 21,
  },
]
`;

exports[`fromNucleicSequence AGGCAG fragment 1`] = `
[
  {
    "atoms": {
      "C": 10,
      "H": 13,
      "N": 5,
      "O": 4,
      "P": 0,
    },
    "charge": 0,
    "comment": "y1",
    "em": 267.09675391942,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C10H13N5O4",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 267.2417245245797,
    "parts": [
      "O-2P-1DgmpH",
    ],
    "sequence": "G",
    "unsaturation": 7,
  },
  {
    "atoms": {
      "C": 20,
      "H": 25,
      "N": 10,
      "O": 9,
      "P": 1,
    },
    "charge": 0,
    "comment": "y2",
    "em": 580.1543594246,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C20H25N10O9P",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 580.4486752178418,
    "parts": [
      "O-2P-1DampDgmpH",
    ],
    "sequence": "AG",
    "unsaturation": 14,
  },
  {
    "atoms": {
      "C": 29,
      "H": 37,
      "N": 13,
      "O": 15,
      "P": 2,
    },
    "charge": 0,
    "comment": "y3",
    "em": 869.20073154049,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C29H37N13O15P2",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 869.6308885157955,
    "parts": [
      "O-2P-1DcmpDampDgmpH",
    ],
    "sequence": "CAG",
    "unsaturation": 19,
  },
  {
    "atoms": {
      "C": 39,
      "H": 49,
      "N": 18,
      "O": 21,
      "P": 3,
    },
    "charge": 0,
    "comment": "y4",
    "em": 1198.25325166524,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C39H49N18O21P3",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 1198.8372441333759,
    "parts": [
      "O-2P-1DgmpDcmpDampDgmpH",
    ],
    "sequence": "GCAG",
    "unsaturation": 26,
  },
  {
    "atoms": {
      "C": 49,
      "H": 61,
      "N": 23,
      "O": 27,
      "P": 4,
    },
    "charge": 0,
    "comment": "y5",
    "em": 1527.3057717899899,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C49H61N23O27P4",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 1528.0435997509564,
    "parts": [
      "O-2P-1DgmpDgmpDcmpDampDgmpH",
    ],
    "sequence": "GGCAG",
    "unsaturation": 33,
  },
  {
    "atoms": {
      "C": 59,
      "H": 74,
      "N": 28,
      "O": 35,
      "P": 6,
    },
    "charge": 0,
    "em": 1920.3297081845299,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C59H74N28O35P6",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 1921.230467969649,
    "parts": [
      "HODampDgmpDgmpDcmpDampDgmpH",
    ],
    "sequence": "AGGCAG",
    "unsaturation": 40,
  },
]
`;

exports[`fromNucleicSequence HODamDamDamDamDamH 1`] = `
[
  {
    "atoms": {
      "C": 12,
      "H": 17,
      "N": 5,
      "O": 4,
      "P": 0,
    },
    "charge": 0,
    "comment": "y1",
    "em": 295.12805404834,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C12H17N5O4",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 295.29495933427324,
    "parts": [
      "O-2P-1DamH",
    ],
    "sequence": "œ",
    "unsaturation": 7,
  },
  {
    "atoms": {
      "C": 24,
      "H": 33,
      "N": 10,
      "O": 10,
      "P": 1,
    },
    "charge": 0,
    "comment": "y2",
    "em": 652.21187430201,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C24H33N10O10P",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 652.5545497615473,
    "parts": [
      "O-2P-1DamDamH",
    ],
    "sequence": "œœ",
    "unsaturation": 14,
  },
  {
    "atoms": {
      "C": 36,
      "H": 49,
      "N": 15,
      "O": 16,
      "P": 2,
    },
    "charge": 0,
    "comment": "y3",
    "em": 1009.2956945556799,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C36H49N15O16P2",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 1009.8141401888214,
    "parts": [
      "O-2P-1DamDamDamH",
    ],
    "sequence": "œœœ",
    "unsaturation": 21,
  },
  {
    "atoms": {
      "C": 48,
      "H": 65,
      "N": 20,
      "O": 22,
      "P": 3,
    },
    "charge": 0,
    "comment": "y4",
    "em": 1366.37951480935,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C48H65N20O22P3",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 1367.0737306160954,
    "parts": [
      "O-2P-1DamDamDamDamH",
    ],
    "sequence": "œœœœ",
    "unsaturation": 28,
  },
  {
    "atoms": {
      "C": 60,
      "H": 82,
      "N": 25,
      "O": 31,
      "P": 5,
    },
    "charge": 0,
    "em": 1803.42966595238,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 0,
      "em": 1.00782503223,
      "mf": "(H+)-",
    },
    "mf": "C60H82N25O31P5",
    "ms": {
      "charge": 0,
      "em": 0,
      "ionization": "(H+)-",
    },
    "mw": 1804.3132385688002,
    "parts": [
      "HODamDamDamDamDamH",
    ],
    "sequence": "œœœœœ",
    "unsaturation": 35,
  },
]
`;

exports[`fromNucleicSequence fromNucleicSequence ds-DNA 1`] = `
[
  {
    "atoms": {
      "C": 10,
      "H": 13,
      "N": 2,
      "O": 7,
      "P": 1,
    },
    "charge": 0,
    "comment": "cmp-a1",
    "em": 304.04603776326,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C10H13N2O7P",
    "ms": {
      "charge": 1,
      "em": 305.05331421558094,
      "ionization": "H+",
    },
    "mw": 304.1935916616171,
    "parts": [
      "HODtmpO-1H-1",
    ],
    "sequence": "T",
    "unsaturation": 6,
  },
  {
    "atoms": {
      "C": 10,
      "H": 12,
      "N": 5,
      "O": 5,
      "P": 1,
    },
    "charge": 0,
    "comment": "a1",
    "em": 313.05760550518,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C10H12N5O5P",
    "ms": {
      "charge": 1,
      "em": 314.0648819575009,
      "ionization": "H+",
    },
    "mw": 313.2069506932622,
    "parts": [
      "HODampO-1H-1",
    ],
    "sequence": "A",
    "unsaturation": 8,
  },
  {
    "atoms": {
      "C": 10,
      "H": 13,
      "N": 2,
      "O": 7,
      "P": 1,
    },
    "charge": 0,
    "comment": "cmp-a1",
    "em": 304.04603776326,
    "ionization": {
      "atoms": {
        "Na": 1,
      },
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C10H13N2O7P",
    "ms": {
      "charge": 1,
      "em": 327.0352584653509,
      "ionization": "Na+",
    },
    "mw": 304.1935916616171,
    "parts": [
      "HODtmpO-1H-1",
    ],
    "sequence": "T",
    "unsaturation": 6,
  },
  {
    "atoms": {
      "C": 10,
      "H": 12,
      "N": 5,
      "O": 5,
      "P": 1,
    },
    "charge": 0,
    "comment": "a1",
    "em": 313.05760550518,
    "ionization": {
      "atoms": {
        "Na": 1,
      },
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C10H12N5O5P",
    "ms": {
      "charge": 1,
      "em": 336.0468262072709,
      "ionization": "Na+",
    },
    "mw": 313.2069506932622,
    "parts": [
      "HODampO-1H-1",
    ],
    "sequence": "A",
    "unsaturation": 8,
  },
  {
    "atoms": {
      "C": 20,
      "H": 26,
      "N": 4,
      "O": 14,
      "P": 2,
    },
    "charge": 0,
    "comment": "cmp-a2",
    "em": 608.09207552652,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C20H26N4O14P2",
    "ms": {
      "charge": 1,
      "em": 609.099351978841,
      "ionization": "H+",
    },
    "mw": 608.3871833232342,
    "parts": [
      "HODtmpDtmpO-1H-1",
    ],
    "sequence": "TT",
    "unsaturation": 11,
  },
  {
    "atoms": {
      "C": 20,
      "H": 24,
      "N": 10,
      "O": 10,
      "P": 2,
    },
    "charge": 0,
    "comment": "a2",
    "em": 626.11521101036,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C20H24N10O10P2",
    "ms": {
      "charge": 1,
      "em": 627.122487462681,
      "ionization": "H+",
    },
    "mw": 626.4139013865243,
    "parts": [
      "HODampDampO-1H-1",
    ],
    "sequence": "AA",
    "unsaturation": 15,
  },
  {
    "atoms": {
      "C": 20,
      "H": 26,
      "N": 4,
      "O": 14,
      "P": 2,
    },
    "charge": 0,
    "comment": "cmp-a2",
    "em": 608.09207552652,
    "ionization": {
      "atoms": {
        "Na": 1,
      },
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C20H26N4O14P2",
    "ms": {
      "charge": 1,
      "em": 631.081296228611,
      "ionization": "Na+",
    },
    "mw": 608.3871833232342,
    "parts": [
      "HODtmpDtmpO-1H-1",
    ],
    "sequence": "TT",
    "unsaturation": 11,
  },
  {
    "atoms": {
      "C": 20,
      "H": 24,
      "N": 10,
      "O": 10,
      "P": 2,
    },
    "charge": 0,
    "comment": "a2",
    "em": 626.11521101036,
    "ionization": {
      "atoms": {
        "Na": 1,
      },
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C20H24N10O10P2",
    "ms": {
      "charge": 1,
      "em": 649.104431712451,
      "ionization": "Na+",
    },
    "mw": 626.4139013865243,
    "parts": [
      "HODampDampO-1H-1",
    ],
    "sequence": "AA",
    "unsaturation": 15,
  },
  {
    "atoms": {
      "C": 30,
      "H": 41,
      "N": 6,
      "O": 22,
      "P": 3,
    },
    "charge": 0,
    "em": 930.14867797381,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C30H41N6O22P3",
    "ms": {
      "charge": 1,
      "em": 931.155954426131,
      "ionization": "H+",
    },
    "mw": 930.5960614172811,
    "parts": [
      "HODtmpDtmpDtmpH",
    ],
    "sequence": "TTT",
    "unsaturation": 15,
  },
  {
    "atoms": {
      "C": 30,
      "H": 41,
      "N": 6,
      "O": 22,
      "P": 3,
    },
    "charge": 0,
    "em": 930.14867797381,
    "ionization": {
      "atoms": {
        "Na": 1,
      },
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C30H41N6O22P3",
    "ms": {
      "charge": 1,
      "em": 953.1378986759009,
      "ionization": "Na+",
    },
    "mw": 930.5960614172811,
    "parts": [
      "HODtmpDtmpDtmpH",
    ],
    "sequence": "TTT",
    "unsaturation": 15,
  },
  {
    "atoms": {
      "C": 30,
      "H": 38,
      "N": 15,
      "O": 16,
      "P": 3,
    },
    "charge": 0,
    "em": 957.18338119957,
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mf": "C30H38N15O16P3",
    "ms": {
      "charge": 1,
      "em": 958.190657651891,
      "ionization": "H+",
    },
    "mw": 957.6361385122163,
    "parts": [
      "HODampDampDampH",
    ],
    "sequence": "AAA",
    "unsaturation": 21,
  },
  {
    "atoms": {
      "C": 30,
      "H": 38,
      "N": 15,
      "O": 16,
      "P": 3,
    },
    "charge": 0,
    "em": 957.18338119957,
    "ionization": {
      "atoms": {
        "Na": 1,
      },
      "charge": 1,
      "em": 22.989769282,
      "mf": "Na+",
    },
    "mf": "C30H38N15O16P3",
    "ms": {
      "charge": 1,
      "em": 980.172601901661,
      "ionization": "Na+",
    },
    "mw": 957.6361385122163,
    "parts": [
      "HODampDampDampH",
    ],
    "sequence": "AAA",
    "unsaturation": 21,
  },
]
`;
